harvester/harvest.py

```diff
@@ -0,0 +1,497 @@
+            if self.extract_type == "waf":
+                i, h = self.get_title_from_fgdc(record["content"]), dataset_to_hash(
+                    record["content"].decode(
+                        "utf-8"
+                    )  # TODO: is this the correct encoding?
+                )
+            self.records[i] = Record(
+                self,
+                i,
+                record,
+                h,
+            )
+
+    def ckan_to_id_hash(self, results: list[dict]) -> None:
+        for record in results:
+            self.ckan_records[record["identifier"]] = dataset_to_hash(
+                eval(
+                    record["dcat_metadata"], {"__builtins__": {}}
+                )  # TODO: eval is a big security risk. think of a better solution.
+            )
+
+    def traverse_waf(self, url, files=[], file_ext=".xml", folder="/", filters=[]):
+        """Transverses WAF
+        Please add docstrings
+        """
+        # TODO: add exception handling
+        parent = os.path.dirname(url.rstrip("/"))
+
+        res = requests.get(url)
+        if res.status_code == 200:
+            soup = BeautifulSoup(res.content, "html.parser")
+            anchors = soup.find_all("a", href=True)
+
+            folders = []
+            for anchor in anchors:
+                if (
+                    anchor["href"].endswith(folder)
+                    and not parent.endswith(anchor["href"].rstrip("/"))
+                    and anchor["href"] not in filters
+                ):
+                    folders.append(os.path.join(url, anchor["href"]))
+
+                if anchor["href"].endswith(file_ext):
+                    # TODO: just download the file here
+                    # instead of storing them and returning at the end.
+                    files.append(os.path.join(url, anchor["href"]))
+
+        for folder in folders:
+            self.traverse_waf(folder, files=files, filters=filters)
+
+        return files
+
+    def download_waf(self, files):
+        """Downloads WAF
+        Please add docstrings
+        """
+        output = []
+        for file in files:
+            res = requests.get(file)
+            if res.status_code == 200:
+                output.append({"url": file, "content": res.content})
+
+        return output
+
+    def compare(self) -> None:
+        """Compares records"""
+        logger.info("Hello from harvester.compare()")
+
+        harvest_ids = set(self.records.keys())
+        ckan_ids = set(self.ckan_records.keys())
+        same_ids = harvest_ids & ckan_ids
+
+        # since python 3.7 dicts are insertion ordered so deletions will occur first
+        self.compare_data = {"delete": None, "create": None, "update": None}
+        self.compare_data["delete"] = ckan_ids - harvest_ids
+        self.compare_data["create"] = harvest_ids - ckan_ids
+        self.compare_data["update"] = set(
+            [
+                i
```

Comment:
list comprehension is easier for me to read on a single line.

Reply:
resolved in d65bed0e328133c937957b5d499cbf5ce0968bb6